src/compact.rs

```diff
@@ -62,8 +62,11 @@ impl<'a, T: 'a + Input> Input for PrefixInput<'a, T> {
 	}
 
 	fn read(&mut self, buffer: &mut [u8]) -> Result<(), Error> {
+		if buffer.is_empty() {
+			return Ok(());
+		}
 		match self.prefix.take() {
-			Some(v) if !buffer.is_empty() => {
+			Some(v) => {
```

Comment:
If the subsequent read fails, then it cannot be retried, since the `prefix` was already moved out.

I guess you could put it back in the error case, not sure what is expected behaviour for this?

Reply:
The Input trait doesn't say anything about providing guarantees  of correct subsequent reads after a failure. As I can see `IoReader` has a similar behavior as well, since it relies on `read_exact` which is "If this function returns an error, it is unspecified how many bytes it has read..."